yt/utilities/lib/particle_kdtree_tools.pyx

```diff
@@ -96,30 +95,203 @@ def generate_smoothing_length(np.float64_t[:, ::1] input_positions,
+            distance (voxels_x, voxels_y, voxels_z, n_neighbors)
+    dists:  Array of floats to store the distance of a nearest neighbor distance
+            (voxels_x, voxels_y, voxels_z, n_neighbors)
+    kdtree: A PyKDTree instance
+    bounds: The region for the voxels to cover in the format x_min, xmax, ymin,
+            ... (6)
+    size:   The number of voxels to have in each dimension
+    input_pos: Array of floats with shape (n_particles, 3)
+            The positions of particles in kdtree sorted order. Currently assumed
+            to be 3D postions.
+    gather_type: An integer with 0 meaning that all dimensions are used in the
+            distance and 1 means only the first two are used. This would be 1
+            in a slice plot.
+    offset: This is the number of gas particle previously calculated and this is
+            to identify a particle from the tree id
+
+    Returns
+    -------
+    No returns, the function mutates the dists and the pids
+
+    """
+    cdef KDTree* c_tree = kdtree._tree
+    cdef Node* leafnode
+    cdef np.float64_t* pos
+    cdef int i, j, k, p
+    cdef BoundedPriorityQueue queue = BoundedPriorityQueue(pids.shape[3], True)
+    cdef np.float64_t dx, dy, dz
+    cdef np.float64_t[::1] voxel_position = np.zeros(shape=(3))
+    cdef np.int64_t[:] tree_id
+
+    tree_id = kdtree.idx.astype("int64")
+
+    dx = (bounds[1] - bounds[0]) / size[0]
+    dy = (bounds[3] - bounds[2]) / size[1]
+    dz = (bounds[5] - bounds[4]) / size[2]
+
+    with nogil:
+        for i in range(size[0]):
+            for j in range(size[1]):
+                for k in range(size[2]):
+                    voxel_position[0] = bounds[0] + (i+0.5)*dx
+                    voxel_position[1] = bounds[2] + (j+0.5)*dy
+                    voxel_position[2] = bounds[4] + (k+0.5)*dz
+
+                    queue.heap[:] = dists[i, j, k, :]
+                    queue.pids[:] = pids[i, j, k, :]
+                    queue.size = pids.shape[3]
+
+                    pos = &(voxel_position[0])
+                    leafnode = c_tree.search(pos)
+
+                    # Traverse the rest of the kdtree to finish the neighbor
+                    # list
+                    find_knn_voxels(c_tree.root, queue, input_pos, offset, tree_id,
+                                pos, leafnode.leafid)
+
+                    dists[i, j, k, :] = queue.heap[:]
+                    pids[i, j, k, :] = queue.pids[:]
+
+@cython.boundscheck(False)
+@cython.wraparound(True)
+@cython.cdivision(True)
+def generate_nn_list_guess(np.int64_t[:, :, :, ::1] pids, np.float64_t[:, :, :, ::1] dists,
+                     np.int64_t[:, : , :] queue_sizes, PyKDTree kdtree,
+                     np.float64_t[:] bounds, np.int64_t[:] size,
+                     np.float64_t[:,::1] input_pos, int offset=0):
+    """Calculate array of distances to the nth nearest neighbor from the
+    leafnode the voxel centre is in, i.e, the best guess of neighbors
```

Comment:
can you not refer explicitly to voxels here and in the implementation? Since this could just be any arbitrary position, not necessarily anything to do with voxels (at least I think, let me know if that's wrong).